helm: add support for history-max parameter

vvanouytsel opened the pull request:
##### SUMMARY
The --history-max parameter allows the user to set a maximum amount of
revisions per release to be kept in the history.

By default helm keeps 10 revisions per release, which means that 10
secrets will be kept per release. If you have multiple helm releases, the amount of secrets created by Helm will quickly add up.

With supporting the 'history-max' CLI parameter via the ansible module, the user will have the posibility to determine how many revisions he wants to keep per release.

##### ISSUE TYPE
- Feature Pull Request


##### COMPONENT NAME
plugins/modules/helm.py

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
**Before change:**
```yaml
- name: example
  hosts: localhost
  tasks:
    - name: "Deploy helm chart before change"
      kubernetes.core.helm:
        name: "example"
        chart_ref: "/tmp/example"
        release_namespace: default
        wait_timeout: 10m
```
```
PLAY [example] ***********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Deploy helm chart before change] ***********************************************************************************************************************************************************************************************
changed: [localhost] => {"changed": true, "command": "/usr/bin/helm upgrade -i --reset-values example /tmp/example", "status": {"app_version": "3.0.0", "chart": "example-3.0.0", "name": "example", "namespace": "default", "revision": "1", "status": "deployed", "updated": "2021-07-08 08:38:47.591116754 +0000 UTC", "values": {}}, "stderr": "", "stderr_lines": [], "stdout": "Release \"example\" does not exist. Installing it now.\nNAME: example\nLAST DEPLOYED: Thu Jul  8 08:38:47 2021\nNAMESPACE: default\nSTATUS: deployed\nREVISION: 1\nTEST SUITE: None\nNOTES:\nDeployed!\n", "stdout_lines": ["Release \"example\" does not exist. Installing it now.", "NAME: example", "LAST DEPLOYED: Thu Jul  8 08:38:47 2021", "NAMESPACE: default", "STATUS: deployed", "REVISION: 1", "TEST SUITE: None", "NOTES:", "Deployed!"]}

PLAY RECAP ***************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

**After change:**
Without history_max  set:
```yaml
- name: example
  hosts: localhost
  tasks:
    - name: "Deploy helm chart without history_max"
      kubernetes.core.helm:
        name: "example"
        chart_ref: "/tmp/example"
        release_namespace: default
        wait_timeout: 10m
```
```
PLAY [example] ***********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Deploy helm chart without history_max] *****************************************************************************************************************************************************************************************
changed: [localhost] => {"changed": true, "command": "/usr/bin/helm upgrade -i --reset-values --history-max=10 example /tmp/example", "status": {"app_version": "3.0.0", "chart": "example-3.0.0", "name": "example", "namespace": "default", "revision": "1", "status": "deployed", "updated": "2021-07-08 08:33:24.728443982 +0000 UTC", "values": {}}, "stderr": "", "stderr_lines": [], "stdout": "Release \"example\" does not exist. Installing it now.\nNAME: example\nLAST DEPLOYED: Thu Jul  8 08:33:24 2021\nNAMESPACE: default\nSTATUS: deployed\nREVISION: 1\nTEST SUITE: None\nNOTES:\nDeployed!\n", "stdout_lines": ["Release \"example\" does not exist. Installing it now.", "NAME: example", "LAST DEPLOYED: Thu Jul  8 08:33:24 2021", "NAMESPACE: default", "STATUS: deployed", "REVISION: 1", "TEST SUITE: None", "NOTES:", "Deployed!"]}

PLAY RECAP ***************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0 
 ```

With history_max set:
```yaml
- name: example
  hosts: localhost
  tasks:
    - name: "Deploy helm chart with history_max"
      kubernetes.core.helm:
        name: "example"
        chart_ref: "/tmp/example"
        release_namespace: default
        wait_timeout: 10m
        history_max: 3
```
```
PLAY [example] ***********************************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [Deploy helm chart with history_max] ********************************************************************************************************************************************************************************************
changed: [localhost] => {"changed": true, "command": "/usr/bin/helm upgrade -i --reset-values --history-max=3 example /tmp/example", "status": {"app_version": "3.0.0", "chart": "example-3.0.0", "name": "example", "namespace": "default", "revision": "1", "status": "deployed", "updated": "2021-07-08 08:35:28.231751834 +0000 UTC", "values": {}}, "stderr": "", "stderr_lines": [], "stdout": "Release \"example\" does not exist. Installing it now.\nNAME: example\nLAST DEPLOYED: Thu Jul  8 08:35:28 2021\nNAMESPACE: default\nSTATUS: deployed\nREVISION: 1\nTEST SUITE: None\nNOTES:\nDeployed!\n", "stdout_lines": ["Release \"example\" does not exist. Installing it now.", "NAME: example", "LAST DEPLOYED: Thu Jul  8 08:35:28 2021", "NAMESPACE: default", "STATUS: deployed", "REVISION: 1", "TEST SUITE: None", "NOTES:", "Deployed!"]}

PLAY RECAP ***************************************************************************************************************************************************************************************************************************
localhost                  : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0  
```

Akasurde commented on plugins/modules/helm.py:
```diff
@@ -139,6 +139,12 @@
+      - Limit the maximum number of revisions saved per release.
+    type: int
+    default: False
+    version_added: "2.1.2"
```
```suggestion
    version_added: "2.2.0"
```

This will go in minor release rather than in patch release.

Akasurde commented on plugins/modules/helm.py:
```diff
@@ -404,7 +410,7 @@ def deploy(command, release_name, release_values, chart_name, wait,
         deploy_command += " --skip-crds"
 
-    deploy_command += " " + release_name + " " + chart_name
+    deploy_command +=  " --history-max=" + str(history_max) + " " + release_name + " " + chart_name
```
```suggestion
    if history_max: 
        deploy_command += " --history-max=%s" % str(history_max)

    deploy_command += " " + release_name + " " + chart_name
```

Akasurde commented on plugins/modules/helm.py:
```diff
@@ -139,6 +139,12 @@
+    description:
+      - Limit the maximum number of revisions saved per release.
+    type: int
+    default: False
```
Please create a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment). See [this fragment](https://github.com/ansible/ansible/blob/c2f528b7686c9573aa8493357219e0bbe4416dd3/changelogs/fragments/reboot-change-default-boot-command.yaml) as an example.

vvanouytsel commented:
Thanks for you review @Akasurde , I've added your remarks to the PR.
If you have any more remarks please let them know.

Thanks!

vvanouytsel commented:
Thanks for reviewing @abikouo!
Are you able to approve the workflow for this PR as well?

abikouo commented:
> Thanks for reviewing @abikouo!
> Are you able to approve the workflow for this PR as well?

Should be ok now.

gravesm reviewed (changes requested):
@vvanouytsel Thank you for the PR! The `--history-max` option is not supported for `helm install` which is used when the `replace` param is set. The `history_max` and `replace` parameters will need to be set to mutually exclusive. This can be done here: https://github.com/ansible-collections/kubernetes.core/blob/b32e5cd06fa90d2d27df3008a421b46fd7983942/plugins/modules/helm.py#L554 and also by updating the documentation for each parameter, an example of which is here: https://github.com/ansible-collections/kubernetes.core/blob/b32e5cd06fa90d2d27df3008a421b46fd7983942/plugins/modules/k8s.py#L65

vvanouytsel commented:
@gravesm Thanks for your review!
I have fixed the linting issue and have also added both 'replace' and 'history_max' to the 'mutually_exclusive' parameter.
Let me know if you require any more changes.

gravesm reviewed (changes requested):
A few small nits and then everything else looks good.

gravesm reviewed (approved):
Thanks @vvanouytsel!
